### ISSUE TITLE
[BUG] Harvester VM Metrics Are Not Accurate

### ISSUE THREAD

**Issue by deiberts86** [harvester/harvester#7086]
**Describe the bug**
Prometheus VM Metrics are not accurate whatsoever where metrics at a percentage level are exceeding 200000% within the Harvester UI VM metrics. 

**To Reproduce**
1. Install Harvester Monitoring Add-On.
2. Create a few VMs and do some workloads.
3. Go to your "Dashboard > Virtual Machine Metrics"
4. Observed excessive percentages.

**Expected behavior**
Percentages shouldn't exceed 100% utilization

**Support bundle**
I'll send one if needed

**Environment**
OS: Harvester version 1.4.0 (FRESH) and tested with Harvester upgrade 1.3.2 - 1.4.0 (UPGRADE)
Hardware: Dell PowerEdge R640 (8-bay)

**Additional context**
Screenshot attached of the VM Metrics
![crazy-metrics-harvestesr](https://github.com/user-attachments/assets/59c758b3-c709-49ce-a321-626dcdf40c3c)

**Comment by CamrynCarter** [comment]
SURE-9477

**Comment by deiberts86** [comment]
[supportbundle_63682b4f-c84a-4f8c-9b23-8f489add4777_2024-12-05T17-12-05Z.zip](https://github.com/user-attachments/files/18027199/supportbundle_63682b4f-c84a-4f8c-9b23-8f489add4777_2024-12-05T17-12-05Z.zip)

**Comment by noahgildersleeve** [comment]
This was reproduced in our lab as well. 

![image](https://github.com/user-attachments/assets/fb7a8b3a-67ea-4e24-98a4-58fbb2287370)
[supportbundle_37afb7f8-8815-43f9-a3eb-56e59d576053_2024-12-19T19-06-54Z.zip](https://github.com/user-attachments/files/18202237/supportbundle_37afb7f8-8815-43f9-a3eb-56e59d576053_2024-12-19T19-06-54Z.zip)

**Comment by w13915984028** [comment]
Seems something is broken, let me take a look, thanks.

**Comment by w13915984028** [comment]
I got this crazy result as well

![image](https://github.com/user-attachments/assets/27c3985f-3664-4bd5-b0d3-b5c684435370)

The `yaxes` needs to be adjusted to a trully percentage data.

https://github.com/harvester/harvester-installer/blob/4f6e6895aabdf491c1ca0e1aed54073879af436f/pkg/config/templates/rancherd-12-monitoring-dashboard.yaml#L127

```
            "targets": [
              {
                "expr": "topk(${count}, (avg(rate(kubevirt_vmi_vcpu_seconds_total[5m])) by (domain, name))) ",
                "instant": false,
                "interval": "",
                "legendFormat": "{{name}}",
                "refId": "A"
              }
            ],
            "thresholds": [],
            "timeFrom": null,
            "timeRegions": [],
            "timeShift": null,
            "title": "CPU Usage",
            "tooltip": {
              "shared": true,
              "sort": 0,
              "value_type": "individual"
            },
            "type": "graph",
            "xaxis": {
              "buckets": null,
              "mode": "time",
              "name": null,
              "show": true,
              "values": []
            },
            "yaxes": [
              {
                "$$hashKey": "object:370",
                "format": "percentunit",
                "label": null,
                "logBase": 1,
                "max": null,
                "min": null,
                "show": true
              },
```

By contrast, the cluster `CPU Utilization` is:
```
...
          "targets": [
            {
              "expr": "1 - avg(irate({__name__=~\"node_cpu_seconds_total|windows_cpu_time_total\",mode=\"idle\"}[$__rate_interval])) by (instance)",
              "interval": "",
              "legendFormat": "{{instance}}",
              "refId": "A"
            }
          ],
          "thresholds": [],
          "timeFrom": null,
          "timeRegions": [],
          "timeShift": null,
          "title": "CPU Utilization",
...
```

**Comment by brandboat** [comment]
Hi @CamrynCarter, @deiberts86, @noahgildersleeve,

As @w13915984028 pointed out, the root cause is due to the incorrect unit. A quick fix is to edit the Grafana dashboard configuration through the Kubernetes ConfigMap.

1. For the `cattle-dashboards/harvester-vm-dashboard` ConfigMap: Search for the JSON object with `"title": "CPU Usage"`, ~~then update the yaxes format from "percentunit" to "percent".~~ update the expr to `topk(${count}, (avg(rate(kubevirt_vmi_vcpu_seconds_total[5m])) by (domain, name)) / 1000)`.

```json
...
"targets": [                                                                                                                                                                
  {                                                                                                                                                                         
    "expr": "topk(${count}, (avg(rate(kubevirt_vmi_vcpu_seconds_total[5m])) by (domain, name))) ",<-- update this line                                                     
    "instant": false,                                                                                                                                                       
    "interval": "",
    "legendFormat": "{{name}}",
    "refId": "A"                                                                                                                                                            
  }
...
"title": "CPU Usage",
...
```

2. Similarly, for the `cattle-dashboards/harvester-vm-detail-dashboard` ConfigMap: Search for the JSON object with `"title": "CPU Usage"`, ~~and change the "unit" field from "percentunit" to "percent".~~ change the expr to `sum(avg(rate(kubevirt_vmi_vcpu_seconds_total{namespace=\"$namespace\", name=\"$vm\"}[5m])) by (domain, name)) / 1000`

```json
      "targets": [
        {
          // update the expr
          "expr": "sum(avg(rate(kubevirt_vmi_vcpu_seconds_total{namespace=\"$namespace\", name=\"$vm\"}[5m])) by (domain, name))",
          "format": "time_series",
          "instant": false,
          "interval": "",
          "intervalFactor": 2,
          "legendFormat": "{{name}}",
          "legendLink": null,
          "refId": "A",
          "step": 10
        }
      ],
      "timeFrom": null,
      "timeShift": null,
      "title": "CPU Usage",
```

This may take a few seconds for the changes to take effect.

**Comment by albinsun** [comment]
Test FAIL, have no crazy number like 4XXXX% before, but still not a reasonable number. It shows 1000% when CPU full loading.
Please help to check.



## Environment
* `harvester-v1.5.0-dev-20250203`
  * Profile: 2 nodes AMD64 QEMU/KVM
  * ui-source: `Auto`

## Prerequisites
1. Create image _ubuntu-22.04-server-cloudimg-amd64.img_
1. Create VM network _mgmt-vlan1_


## Test Cases
1. <details>
   <summary> Enable <b>rancher-monitoring</b> Add-on
   </summary>
   
   * Add-on
     ![Image](https://github.com/user-attachments/assets/b248f6d4-c650-4357-b1ea-e400a86a5986)
   * Metric
     ![Image](https://github.com/user-attachments/assets/e8bcf0c9-d446-4d6a-8316-d3c664d7f8b6)
   </details>

1. Create some VMs and having burn some workloads 
   * Leverage **stress** util (`sudo apt-get install stress`)
   * 3 VMs: _1cpu_, _2cpu_ and _4cpu_, and burn CPU on _1cpu_ and _2cpu_.
   <br/>

1. <details open>
   <summary> 🔴 Go <b>Dashboard > Virtual Machine Metrics</b>, check CPU Usage
   </summary>

   > ⚠ Have no crazy number like 4XXXX% before, but the amout still seems not correct. It shows 1000% when full loading.
   ![Image](https://github.com/user-attachments/assets/f6e7d2f3-3de1-43d9-b3b5-b3004332e895)
   </details>

1. <details open>
   <summary> 🔴 Turn off CPU burning on <i>2cpu</i>, keeping on <i>1cpu</i>
   </summary>
   
   ![Image](https://github.com/user-attachments/assets/2e139f06-6a18-4e75-9184-4b1fc864930d)
   </details>


## Logs
* Support bundle: [supportbundle_grafana-cpu-usage_2025-02-04T04-26-45Z.zip](https://github.com/user-attachments/files/18651689/supportbundle_grafana-cpu-usage_2025-02-04T04-26-45Z.zip)

**Comment by brandboat** [comment]
~~The current CPU usage formula seems weird. I think if a VM has 2 CPUs, and each CPU is at 100% usage, the percentage should be 200%, not 1000% as shown on the dashboard (as mentioned in https://github.com/harvester/harvester/issues/7086#issuecomment-2632801527). I'm thinking maybe we should change the CPU usage formula to `rate(kubevirt_vmi_cpu_usage_seconds_total[5m])`. c.c. @w13915984028, what do you think?~~

Update: kubevirt_vmi_cpu_usage_seconds_total = (sum of both vcpu + hypervisor usage), and we don't want the part of hypervisor usage, so it is not suitable for showing vm cpu usage here.
https://kubevirt.io/monitoring/metrics.html#kubevirt_vmi_cpu_usage_seconds_total

**Comment by brandboat** [comment]
There are 2 issues in the current cpu usage formula `sum(avg(rate(kubevirt_vmi_vcpu_seconds_total{namespace=\"$namespace\", name=\"$vm\"}[5m])) by (domain, name))`:
- `avg(rate(...)) by (domain, name)` average each vcpu usage to one value, e.g. if `rate(kubevirt_vmi_vcpu_seconds_total...[5m])` is 1000 in both 2 vcpus, the output will be (1000 + 1000) / 2 = 1000, and what we expect is 2000, we should remove the avg out of the formula.
- The unit of `kubevirt_vmi_vcpu_seconds_total` seems wrong https://github.com/kubevirt/kubevirt/issues/13731, if cpu usage is 100%, the value of `kubevirt_vmi_vcpu_seconds_total` should be 1 instead of 1000.

In short, we need to update the formula to: `sum(rate(kubevirt_vmi_vcpu_seconds_total{namespace="default", name="cpu2"}[5m])) / 1000`, and change the y-axis unit to `percentunit` in Harvester VM Detail Dashboard. And Harvester VM Dashboard should do the similar things.

c.c. @bk201, @w13915984028

**Comment by albinsun** [comment]
> There are 2 issues in the current cpu usage formula
> ...
> * `avg(rate(...)) by (domain, name)` average each vcpu usage to one value, e.g. if `rate(kubevirt_vmi_vcpu_seconds_total...[5m])` is 1000 in both 2 vcpus, the output will be (1000 + 1000) / 2 = 1000, and what we expect is 2000, we should remove the avg out of the formula.

Except this absolute usage proposal, may also consider to keep current averaged approach.
Consider 2 VMs,
1. _cpu1_ w/ 1 vCPU : Busy one, averagely in quite high loading, let's say 1000.
2. _cpu8_ w/ 8 vCPUs : Spare one, averagely 4 vCPUs (50%) working.

And we will have different metrics as
* Absolute Approach
  1. _cpu1_: `1000`
  2. _cpu8_: `4000`
* Averaged Approach (Current)
  1. _cpu1_: `1000` (100%)
  2. _cpu8_: `500` (50%) (4000 / 8)

For me, the averaged approach seems more intuitive cuz user can quickly know _cpu8_ looks fine but _cpu1_ may need more vCPU as it averagely in high loading.

**Comment by brandboat** [comment]
> For me, the averaged approach seems more intuitive cuz user can quickly know cpu8 looks fine but cpu1 may need more vCPU as it averagely in high loading.

@albinsun, thanks. That makes sense to me, perhaps that's the reason why we add `avg` in the prometheus query. This issue should focus only on fixing the incorrect metric values shown on the Grafana dashboard (i.e., dividing by 1000), rather than changing the entire prometheus query. If we want to change the prometheus query, it might be worth discussing in a separate issue.

**Comment by albinsun** [comment]
Test FAIL, number looks great in individual figures, but seems some misplace/miscalculated in the overall figure, Please help to check.



## Environment
* `harvester-v1.5-66ff5df3-head`
  * Profile: 3 nodes witness AMD64 QEMU/KVM
  * ui-source: `Auto`

## Logs
* Support bundle: [supportbundle_cpuusage_2025-02-14T05-16-48Z.zip](https://github.com/user-attachments/files/18793858/supportbundle_cpuusage_2025-02-14T05-16-48Z.zip)

## Test Scenario
1. <details>
   <summary> Enable <b>rancher-monitoring</b> Add-on
   </summary>
   
   * Add-on
     ![Image](https://github.com/user-attachments/assets/a1477191-86bb-40f1-a601-3d1c76a25d41)
   * Metric
     ![Image](https://github.com/user-attachments/assets/5cd3d65f-b945-42c6-8d34-aee3ce33f00a)
   </details>

1. <details>
   <summary> Create 3 VMs, check <b>Dashboard > Virtual Machine Metrics > CPU Usage</b>, should be ~0%
   </summary>
   
   * VM
     ![Image](https://github.com/user-attachments/assets/ea2227dd-2caa-4c8e-b73c-146cd8d7910b)
   * CPU Usage
     ![Image](https://github.com/user-attachments/assets/facb8591-dd11-4a85-8c2b-3573feb31b40)
   </details>

   > Install **stress** util (`sudo apt-get install stress`)
   > 3 VMs: _1cpu_, _2cpu_ and _4cpu_

1. <details open>
   <summary> Burn 1 cpu on both <i>1cpu</i> and <i>2cpu</i>, check <b>CPU Usage</b>, <i>1cpu</i> / <i>2cpu</i> / <i>4cpu</i> should be ~100% / ~50% / ~25%
   </summary>

   * Burn CPU
     ![Image](https://github.com/user-attachments/assets/a14fdb89-09bb-4c1d-abe2-32630e412889)

   * 🔴 Overall
     > 1cpu should be ~100% but not there
     > 4cpu should be ~25% but shows 150%
     ![Image](https://github.com/user-attachments/assets/a09c3623-5732-405d-8db9-8b74ff5aca7b)

   * 🟢 individually, 1cpu should be ~100%
     ![Image](https://github.com/user-attachments/assets/5b2ae4c6-ee26-4251-89bb-fdb210e1333f)

   * 🟢 individually, 2cpu should be ~50%
     ![Image](https://github.com/user-attachments/assets/6172447c-f2c1-46cb-9d29-c9586a4fd037)

   * 🟢 individually, 4cpu should be ~25%
     ![Image](https://github.com/user-attachments/assets/45473f52-7256-4650-ad97-3ac5e9548143)
   </details>

**Comment by brandboat** [comment]
> 4cpu should be ~0% but shows 150%

@albinsun, this is an expected behavior since we are using `stack series` in these panels. And the metrics also show 4cpu is 0.169%, more info please check https://grafana.com/docs/grafana/latest/panels-visualizations/visualizations/time-series/#stack-series.

**Comment by albinsun** [comment]
Thanks @brandboat, that makes sense to me, let's close as fixed.
In short, the overall figure is a stacked view and we can get individual values in the hovering picture.

**Comment by innobead** [comment]
Added the right backport label for 1.4.2 instead of 1.4.3.
